File: python/breakout/ball.py
import pygame

class Ball():
    def __init__(self, screen, bar, settings, chances):
        self.settings = settings
        self.bar = bar
        self.screen = screen
        self.chances = chances

        #cor branca
        self.color = (255, 255, 255)

        #especificações da bolinha
        self.radius = 15
        self.center = [self.bar.rect.centerx, self.bar.rect.top - self.radius]

        #forma o retângulo para a bolinha (necessário para detectar colisôes com a barra)
        self.rect = pygame.Rect(0, 0, 2*self.radius, 2*self.radius)

        self.deflection = 0

        #flags de movimentação
        self.up = False
        self.down = False
        self.left = False
        self.right = False

    def draw(self):
        pygame.draw.circle(self.screen, self.color, self.center, self.radius)

    def update(self):
        if not self.settings.game_start:
            self.follow_bar()
        else:
            self.set_flags()
            self.move()

    def set_flags(self):
        #movimento vertical
        if (self.center[1] - self.radius) > 0 and not self.down:
            self.go_up()
        else:
            self.go_down()
            if (self.center[1] + self.radius) >= self.screen.get_rect().height:
                self.fail()
                self.bar.rect.centerx = self.screen.get_rect().centerx

        #movimento horizontal
        if self.right or self.left:
            if (self.center[0] - self.radius) > 0 and not self.right:
                self.go_left()
            else:
                self.go_right()
                if (self.center[0] + self.radius) >= self.screen.get_rect().width:
                    self.go_left()

    def go_up(self):
        self.up = True
        self.down = False

    def go_down(self):
        self.up = False
        self.down = True

    def go_left(self):
        self.right = False
        self.left = True

    def go_right(self):
        self.left = False
        self.right = True

    def fail(self):
        self.up = False
        self.down = False
        self.right = False
        self.left = False
        self.chances.left -= 1
        self.what_to_do()

    def what_to_do(self):
        if self.chances.check_chances():
            self.settings.pre_start = True
            self.settings.game_start = False
        else:
            self.settings.pre_start = False
            self.settings.game_start = False
            self.settings.game_over = True

    def follow_bar(self):
        self.center[0] = self.bar.rect.centerx
        self.center[1] = self.bar.rect.top - self.radius

    def move(self):
        if self.up:
            self.center[1] -= 1

        if self.down:
            self.center[1] += 1

        if self.left:
            self.center[0] -= self.deflection

        if self.right:
            self.center[0] += self.deflection

        self.update_rect()

    def update_rect(self):
        self.rect.centerx = self.center[0]
        self.rect.centery = self.center[1]

    def direction(self, centerx):
        #direita
        if centerx > self.bar.rect.centerx + 25 and centerx <= self.bar.rect.centerx + 40:
            self.deflection = 0.1
            self.right = True
            self.left = False

        elif centerx > self.bar.rect.centerx + 40 and centerx <= self.bar.rect.centerx + 60:
            self.deflection = 0.2
            self.right = True
            self.left = False

        elif centerx > self.bar.rect.centerx + 60 and centerx <= self.bar.rect.centerx + 85:
            self.deflection = 0.3
            self.right = True
            self.left = False

        elif centerx > self.bar.rect.centerx + 85:
            self.deflection = 0.5
            self.right = True
            self.left = False

        #esquerda
        elif centerx < self.bar.rect.centerx - 25 and centerx >= self.bar.rect.centerx - 40:
            self.deflection = 0.1
            self.right = False
            self.left = True

        elif centerx < self.bar.rect.centerx - 40 and centerx >= self.bar.rect.centerx - 60:
            self.deflection = 0.2
            self.right = False
            self.left = True

        elif centerx < self.bar.rect.centerx - 60 and centerx >= self.bar.rect.centerx - 85:
            self.deflection = 0.3
            self.right = False
            self.left = True

        elif centerx < self.bar.rect.centerx - 85:
            self.deflection = 0.5
            self.right = False
            self.left = True

File: java/BreakOut/src/breakout/Ball.java
package breakout;

import java.awt.Graphics2D;
import java.awt.Rectangle;

public final class Ball {
    Settings settings;
    
    //position
    private final int startx;
    private final int starty;
    private final int radius;
    private int verticalValue;
    private int horizontalValue;
    private int center;
    
    //movement flags
    private boolean up;
    private boolean down;
    private boolean left;
    private boolean right;
    private boolean allow;
    
    protected final Rectangle rect;
    private final Bar bar;
    private final Score score;
    
    public Ball(Settings set, Bar ba, Score scor){
        settings = set;
        bar = ba;
        score = scor;
        
        //position
        startx = 480;
        starty = 506;
        radius = 20;
        verticalValue = 8;
        
        //movement flags
        resetFlags();
        
        //ball rect
        rect = new Rectangle(startx, starty, 2*radius, 2*radius);
    }
    
    protected void execute(Graphics2D g){
        draw(g);
        if(settings.start){
            checkEdges();
            move();
        }
    }
    
    private void draw(Graphics2D g){
        center = rect.x + (rect.width/2);
        g.fillArc(rect.x, rect.y, 2*radius, 2*radius, 0, 360);
    }
    
    private void checkEdges(){
        //movimento vertical
        if(rect.y > 0 && !(down)){
            goUp();
        }else{
            goDown();
            if ((rect.y + 2*radius) >= settings.screenHeight) reset("fail");    
        }
        
        //movimento horizontal
        if(rect.x > 0 && !(right) && allow){
            goLeft();
        }else if(allow){
            goRight();
            if((rect.x + 2*radius) >= settings.screenWidth) goLeft();
        }
    }
    
    private void move(){
        if(up) rect.y -= verticalValue;
        
        else if(down) rect.y += verticalValue;
        
        if(right) rect.x += horizontalValue;
        
        else if(left) rect.x -= horizontalValue;
    }
    
    protected void reset(String cause){
        rect.x = startx;
        rect.y = starty;
        resetFlags();
        bar.reset();
        
        switch(cause){
            case "fail":
                score.chancesLeft--;
                if(score.chancesLeft == 0) {
                    settings.preStart = false;
                    settings.gameOver = true;
                    verticalValue = 8;
                }
                break;
                
            case "new level":
                verticalValue += 1;
                break;
            
            default:
                break;
        }
        
    }
    
    protected void resetFlags(){
        up = false;
        down = false;
        right = false;
        left = false;
        allow = false;
        settings.preStart = true;
        settings.start = false;
    }
    
    protected void changeDirection(){
        if(down) goUp();
        else if(up) goDown();
    }
    
    protected void touchBar(int centerBar){
      if(center > centerBar){
          if(center < (centerBar+40)) horizontalValue = 1;
          else if(center < (centerBar+75)) horizontalValue = 2;
          else if(center < (centerBar+100)) horizontalValue = 3;
          allow = true;
          goRight();
      }else if (center < centerBar){
          if(center > (centerBar-40)) horizontalValue = 1;
          else if(center > (centerBar-75)) horizontalValue = 2;
          else if(center > (centerBar-100)) horizontalValue = 3;
          allow = true;
          goLeft();
      }
      goUp();
    }
    
    private void goUp(){
        up = true;
        down = false;
    }
    
    private void goDown(){
        up = false;
        down = true;
    }
    
    private void goLeft(){
        left = true;
        right = false;
    }
    
    private void goRight(){
        left = false;
        right = true;
    }
}
